### ISSUE TITLE
ydlee - Users blocked by `blockClaim` can still make claims.

### ISSUE THREAD

**Issue by sherlock-admin2** [sherlock-audit/2024-03-zap-protocol-judging#95]
ydlee

high

# Users blocked by `blockClaim` can still make claims.

## Summary
When claiming for a given `TokenSale`, user is checked if he is blocked from claiming. But the checking does not works well, leading to blocked users can still make claims.

## Vulnerability Detail
When claiming for a given `TokenSale`, `admin.blockClaim` is used to check if the user is blocked from claiming. However, the checking is incorrect, as `address(this)` instead of `msg.sender` is passed to `admin.blockClaim`. This leads to blocked user can still claim for the `TokenSale`.
```solidity
364:    function claim() external {
365:        checkingEpoch();
366:        require(
367:@>          uint8(epoch) > 1 && !admin.blockClaim(address(this)),
368:            "TokenSale: Not time or not allowed"
369:        );
```
https://github.com/sherlock-audit/2024-03-zap-protocol/blob/main/zap-contracts-labs/contracts/TokenSale.sol#L364-L369

## Impact
Blocked users can still claim for `TokenSale`.

## Code Snippet
https://github.com/sherlock-audit/2024-03-zap-protocol/blob/main/zap-contracts-labs/contracts/TokenSale.sol#L364-L369

## Tool used

Manual Review

## Recommendation
Use `msg.sender` instead of `address(this)` when checking if user is blocked from claiming.
```solidity
364:    function claim() external {
365:        checkingEpoch();
366:        require(
367:-           uint8(epoch) > 1 && !admin.blockClaim(address(this)),
367:+           uint8(epoch) > 1 && !admin.blockClaim(msg.sender),
368:            "TokenSale: Not time or not allowed"
369:        );
```

Duplicate of #82
